packages/cel/src/adapter/proto.ts

```diff
@@ -198,27 +221,61 @@ export class ProtoValAdapter implements CelValAdapter {
       }
       return this.toCel(unpacked);
     }
-
-    if (isProtoMsg(native) && !isCelMsg(native)) {
-      if (isMessage(native, UInt32ValueSchema)) {
-        return create(UInt64ValueSchema, {
+    if (isMessage(native, UInt32ValueSchema)) {
+      return this.toCel(
+        create(UInt64ValueSchema, {
           value: BigInt(native.value),
-        });
-      }
-      if (isMessage(native, Int32ValueSchema)) {
-        return create(Int64ValueSchema, {
+        }),
+      );
+    }
+    if (isMessage(native, Int32ValueSchema)) {
+      return this.toCel(
+        create(Int64ValueSchema, {
           value: BigInt(native.value),
-        });
-      }
-      if (isMessage(native, FloatValueSchema)) {
-        return create(DoubleValueSchema, {
+        }),
+      );
+    }
+    if (isMessage(native, FloatValueSchema)) {
+      return this.toCel(
+        create(DoubleValueSchema, {
           value: native.value,
-        });
+        }),
+      );
+    }
+    if (isProtoMsg(native)) {
+      native = reflect(this.getSchema(native.$typeName), native);
+    }
+    if (isReflectMessage(native)) {
+      if (
+        isCelMsg(native.message) ||
+        [
+          FloatValueSchema.typeName,
+          Int32ValueSchema.typeName,
+          UInt32ValueSchema.typeName,
+        ].includes(native.desc.typeName)
+      ) {
+        return this.toCel(native.message);
+      }
+      return new CelObject(native, this, this.getMetadata(native.desc).TYPE);
```

Comment:
Took me a long time to figure out that this does preserve CEL messages as the concrete native type 😓 

In fact, a quick test was the fastest way to know for sure: https://github.com/bufbuild/cel-es/pull/144

Going forward, I don't think we can avoid adding basic unit tests for changes like this.